oak-run/src/main/java/org/apache/jackrabbit/oak/plugins/document/check/OrphanedNodeCheck.java

```diff
@@ -0,0 +1,159 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.jackrabbit.oak.plugins.document.check;
+
+import java.io.Closeable;
+import java.io.IOException;
+import java.util.concurrent.BlockingQueue;
+import java.util.concurrent.Callable;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.LinkedBlockingQueue;
+import java.util.concurrent.ThreadPoolExecutor;
+import java.util.concurrent.TimeUnit;
+
+import org.apache.jackrabbit.oak.commons.concurrent.ExecutorCloser;
+import org.apache.jackrabbit.oak.plugins.document.DocumentNodeState;
+import org.apache.jackrabbit.oak.plugins.document.DocumentNodeStore;
+import org.apache.jackrabbit.oak.plugins.document.NodeDocument;
+import org.apache.jackrabbit.oak.plugins.document.Path;
+import org.apache.jackrabbit.oak.plugins.document.RevisionVector;
+import org.apache.jackrabbit.oak.spi.state.NodeState;
+import org.jetbrains.annotations.NotNull;
+
+/**
+ * <code>OrphanedNodeCheck</code>...
+ */
+public class OrphanedNodeCheck implements DocumentProcessor, Closeable {
+
+    private final DocumentNodeStore ns;
+
+    private final RevisionVector headRevision;
+
+    private final ExecutorService executorService;
+
+    public OrphanedNodeCheck(DocumentNodeStore ns,
+                             RevisionVector headRevision,
+                             int numThread) {
+        this.ns = ns;
+        this.headRevision = headRevision;
+        this.executorService = new ThreadPoolExecutor(
+                numThread, numThread, 1, TimeUnit.MINUTES,
+                new LinkedBlockingQueue<>(1000),
+                new ThreadPoolExecutor.CallerRunsPolicy()
+        );
+    }
+
+    @Override
+    public void processDocument(@NotNull NodeDocument document,
+                                @NotNull BlockingQueue<Result> results) {
+        if (!document.isSplitDocument()) {
+            executorService.submit(new CheckDocument(ns, headRevision, document, results));
+        }
+    }
+
+    @Override
+    public void end(@NotNull BlockingQueue<Result> results)
+            throws InterruptedException {
+        executorService.shutdown();
+        if (!executorService.awaitTermination(5, TimeUnit.MINUTES)) {
+            String msg = "Checks still not finished after the last one has been submitted 5 minutes ago";
+            results.put(() -> {
+                StringBuilder sb = new StringBuilder("{");
+                sb.append("\"time\": \"").append(nowAsISO8601()).append('"');
+                sb.append(", \"info\": \"").append(msg).append('"');
+                sb.append("}");
+                return sb.toString();
+            });
+        }
+    }
+
+    @Override
+    public void close() throws IOException {
+        new ExecutorCloser(executorService, 5, TimeUnit.MINUTES).close();
+    }
+
+    private static final class CheckDocument implements Callable<Void> {
```

Comment:
I would change to implement `Runnbale` since we are neither returning anything nor throwing any exceptions.

Reply:
Unfortunately that's not quite true. The task implementation puts a `Result` in the `results` queue, which may throw an `InterruptedException`. I'll leave this as is.